### PR TITLE
Optimize swap operation for Cyclic and Stencil distributions

### PR DESCRIPTION
**Pull request by bradcray**
This PR optimizes the swap operation for arrays distributed using the Cyclic and Stencil distributions, similar to what had been done for Block previously in #16246.  This should benefit @jeremiah-corrado's Stencil-based codes in https://github.com/chapel-lang/chapel/pull/22655 since they rely on the swap operator, and I'm hoping to rely on them to show the improvement in performance rather than adding new swap-specific performance tests here.  In some (lame) tests on my desktop, I saw speedups in the 3/8 (for oversubscribed gasnet stencil codes) to 100x (for a swap microbenchmark), but these were not principled experiments.

The implementation approach was essentially a copy-paste of the Block distribution's logic with the major changes being to the debugging output for Cyclic, which suggests a potential opportunity to re-use code in future refactorings, though I didn't have the time or motivation to try that here.

Most of the other code is new tests which are variants of @e-kayrakli's `edgeCases.chpl` test in the same directory, where the Stencil version is essentially a clone of the Block logic from that, and the Cyclic one tries to generate similar scenarios, but in a Cyclic world.

While here, I found and fixed a bug (that I may have introduced, or at least not fixed previously) in how Stencil distributions print themselves out where the pure low/high bounds were being used rather than the aligned ones, resulting in OOB errors.  Added a test to lock that in along with the Cyclic swap test above.

### CODE REVIEW

**Comment by bradcray** [comment]
@jeremiah-corrado : This extends the optional optimization for swapping arrays that has been implemented for Block arrays to Cyclic and Stencil arrays as well.  In a quick-and-dirty desktop experiment, it resulted in a ~100x speedup for a silly little array swap benchmark I tried.  Within the context of your stencil codes, I'm seeing more like a 2x-10x improvement running 1 locale with GASNet on my desktop.  Don't read anything specific into these numbers other than that it should reduce the gap between block and stencil signficantly (though I wouldn't be at all surprised if there were other things Block does that Stencil could as well).

I'm proposing we hold off on merging this until your tests are merged and have a few nights of performance so that we can see the impact in the nightly perf graphs.

Tagging @bmcdonald3 for his interest, as he implemented this optimization originally.

**Comment by bradcray** [comment]
@e-kayrakli , @bmcdonald3 , @benharsh : Would one of you have time to review this PR (where touch points include the array swap optimization and the Stencil distribution).

**Comment by bmcdonald3** [comment]
Sure, I can give this PR a review today.

File: modules/dists/CyclicDist.chpl
Comment (bmcdonald3) [inline modules/dists/CyclicDist.chpl:1462]: This will always be `true`, right? Is this conditional necessary?

File: modules/dists/StencilDist.chpl
Comment (bmcdonald3) [inline modules/dists/StencilDist.chpl:2231]: Same as above, I am not seeing the point of this conditional.

**Review by bmcdonald3 — APPROVED** [review]
Looks good, just one question about the first check of `domsMatch`.

File: modules/dists/CyclicDist.chpl
Comment (bradcray) [reply modules/dists/CyclicDist.chpl:1462]: Hmm, good point.  TBH, I copy-pasted this from the Block swap routine without much thought, which seems to have the same property.  @e-kayrakli, looks like you added this pattern in https://github.com/chapel-lang/chapel/pull/16620.  Are you seeing anything we're missing?

Comment (e-kayrakli) [reply modules/dists/CyclicDist.chpl:1462]: Yeah, that looks strange. Probably a leftover from an earlier implementation of some kind. Could you drop the equivalent in the Block implementation while here, or do you want me to?

Comment (bradcray) [reply modules/dists/CyclicDist.chpl:1462]: Yep, I've already got it, was just waiting for confirmation and a quick test before committing.  Thanks for taking a look!

**Pull request merged**
